kafkabp/consumer.go

```diff
@@ -0,0 +1,262 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumerMessage represents a kafka consumer message.
+type ConsumerMessage struct {
+	Key, Value []byte
+	Topic      string
+	Partition  int32
+	Offset     int64
+	Timestamp  time.Time
+}
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg     ConsumerConfig
+	topic   string
+	offset  int64
+	tracing bool
+	name    string
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+func (kc *consumer) getConsumer() sarama.Consumer {
+	c, _ := kc.consumer.Load().(sarama.Consumer)
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// Reset recreates the consumer and assigns partitions.
+func (kc *consumer) Reset() error {
```

Comment:
Reset probably shouldn't be exported since it doesn't seem like something you would want people to just be able to call? (granted, it's effectively not exported because `consumer` isn't but it is still probably better from a hygiene/setting a good example perspective)